contributors/devel: add welcome-to-kubernetes-new-developer-guide.md

alexandercampbell opened the pull request:
Add a new 1-page document targeted to brand new developers. Some of this information is duplicated elsewhere but there is no "onboarding" doc for Kubernetes developers (as far as I could discern).

Review welcome!
@grodrigues3 and @pwittrock have already done a couple rounds of review on this in Google docs.

Thanks!

cblecker commented:
@kubernetes/sig-contributor-experience-pr-reviews

grodrigues3 commented:
@castrojo @spzala 

I think both of you were interested in work in this area.

spzala commented on contributors/devel/welcome-to-kubernetes-new-developer-guide.md:
```diff
@@ -0,0 +1,128 @@
+---
+
+_This document assumes that you know what Kubernetes does. If you don't, search
+YouTube for ["Kubernetes"](https://www.youtube.com/watch?v=8SvQqZNP6uo) and try
```
I think you should give this link as one of the examples instead of hyperlink. When we say search for "Kubernetes" in youtube - the correct link will be something like https://www.youtube.com/results?search_query=kubernetes

alexandercampbell replied:
I have updated to change the phrasing but left the link the same. What do you think?

spzala replied:
That looks good to me. Thanks!

spzala commented:
@grodrigues3 thanks for the mention. @alexandercampbell Alex, thanks for putting it together..I have few inline comments.

alexandercampbell commented on contributors/devel/welcome-to-kubernetes-new-developer-guide.md:
```diff
@@ -0,0 +1,128 @@
+able to pick up issues, write code to fix them, and get your work reviewed and
+merged. If you have questions that aren't answered in this document, feel free
+to suggest changes to it directly or email me at
+[alexanderhcampbell@gmail.com](mailto:alexanderhcampbell@gmail.com).
```
I decided to remove this entire sentence since the information is duplicated in a later paragraph

```markdown
There is an entire SIG
([sig-contributor-experience](https://github.com/kubernetes/kubernetes/wiki/CLA-FAQ))
devoted to improving your experience as a contributor. Contributing to
Kubernetes should be easy. If you find a rough edge, let us know!  File an issue
in the appropriate repository if you know it, send a message on [our Slack
channel](https://kubernetes.slack.com/messages/C1TU9EB9S/details) or to the
[owners](https://github.com/kubernetes/community/blob/master/contributors/devel/owners.md)
of contributor-experience. Better yet, help us fix it by joining the SIG; just
show up to one of the [bi-weekly
meetings](https://docs.google.com/document/d/1qf-02B7EOrItQgwXFxgqZ5qjW0mtfu5qkYIF1Hl4ZLI/edit).
```

alexandercampbell commented on contributors/devel/welcome-to-kubernetes-new-developer-guide.md:
```diff
@@ -0,0 +1,128 @@
+repository. You can see an example of this planning process at
+[kubernetes/community#645](https://github.com/kubernetes/community/pull/645). Like everything else in Kubernetes, a SIG is an open,
+community, effort. In each SIG, full-time employees from the
+Kubernetes Partners work side-by-side with volunteer contributors. Anybody is
```
I've updated this paragraph. Let me know what you think!

alexandercampbell commented on contributors/devel/welcome-to-kubernetes-new-developer-guide.md:
```diff
@@ -0,0 +1,128 @@
+ 1. @k8s-bot
+ 2. @k8s-ci-robot
+ 3. @k8s-merge-robot
+
```
I cannot find the usage or work of the @k8s-bit. If you let me know what it does I will add a description to the doc :)

spzala replied:
:) actually me neither and that's why I thought about it. As I look at various PRs and Issues - I only notice @k8s-ci-robot and @k8s-merge-robot in play. Hopefully some reviewer will provide some info, if not I wonder if we should remove it from the list. The https://github.com/kubernetes/test-infra/blob/master/commands.md#k8s-bot-commands provides commands for ci-robot and merge-robot.

cblecker replied:
k8s-bot is an older name. @k8s-ci-robot is the "prow" tool, and @k8s-merge-robot is the "mungegithub" tool. You could also link to the kubernetes/test-infra repo where those live, and the https://git.k8s.io/test-infra/commands.md doc that is a listing of all the current commands and what implements them.

spzala replied:
Thanks to help it understand @cblecker  So I guess in that case let's just remove mentioning k8s-bot!

alexandercampbell commented on contributors/devel/welcome-to-kubernetes-new-developer-guide.md:
```diff
@@ -0,0 +1,128 @@
+file of the your changes' root directory and you should get a quicker response.
+
+In order for your PR to be merged, you need four things. When you have all four,
+the merge robot will automatically merge your commits.
```
Good call. Updated.

alexandercampbell commented:
@spzala thanks for the review! I've responded to all of your comments. I will squash the commits if you like how this looks.

spzala commented on contributors/devel/welcome-to-kubernetes-new-developer-guide.md:
```diff
@@ -0,0 +1,128 @@
+and reviewing code.
+
+Most people who visit the Kubernetes repository for the first time are
+bewildered by the thousands of open issues in our main repository. But now that
```
I am not sure if you forgot to address this comment? Thanks!

alexandercampbell replied:
I did address the comment, but maybe something is up with the way Github is displaying the content. I should have commented a response though. Does it look ok now?

spzala replied:
Yup, see it now, that's good. Thanks!

cblecker commented:
/cc
Will give this a look over today

cblecker commented on contributors/devel/welcome-to-kubernetes-new-developer-guide.md:
```diff
@@ -0,0 +1,126 @@
+============================================
+
+2017-06-21 through 2017-08-06
+Original author: [Alexander Campbell](https://github.com/alexandercampbell)
```
Original author and dates aren't needed as they are a part of git history (I'm guessing this started as a google doc or something?)

alexandercampbell replied:
My justification for putting my name on things: I'm not a fan of the way the docs end up floating around ownerless and unmaintained.

I can remove the dates for sure. What do you think?

cblecker replied:
Just my opinion: original author information is point in time. I don't think it encourages/discourages anyone from participating and updating things. If you want to continue updating this document after it's merged, I think new contributors would be glad to have your contributions. It's really on all of us (the k8s community, and sig-contribex) to update these documents and keep them relevant, not on one person. Perhaps identifying the sig that owns it (like contribex)?

alexandercampbell replied:
Ok, makes sense. I'll remove it. Thanks for the feedback.

cblecker commented on contributors/devel/welcome-to-kubernetes-new-developer-guide.md:
```diff
@@ -0,0 +1,126 @@
+-----------------------
+
+Kubernetes developers work in teams called Special Interest Groups (SIGs). At
+the time of this writing there are 31 SIGs (I got the number 31 by going to the
```
How you got the number probably isn't critical for this doc. I'd also link out to the list of sigs (../../sig-list.md)

cblecker commented on contributors/devel/welcome-to-kubernetes-new-developer-guide.md:
```diff
@@ -0,0 +1,126 @@
+of Kubernetes. SIGs organize themselves by submitting markdown design documents
+to the [kubernetes/community](https://github.com/kubernetes/community)
+repository. You can see an example of this planning process at
+[kubernetes/community#645](https://github.com/kubernetes/community/pull/645). Like everything else in Kubernetes, a SIG is an open,
```
nit: 80 char/line

cblecker commented on contributors/devel/welcome-to-kubernetes-new-developer-guide.md:
```diff
@@ -0,0 +1,126 @@
+bewildered by the thousands of [open
+issues](https://github.com/kubernetes/kubernetes/issues) in our main repository.
+But now that you know about SIGs, it's easy to filter by labels to see what's
+going on in your SIG. For more information about our issue system, check out
```
s/your/a particular sig/
a contributor can be involved in one or more

cblecker commented on contributors/devel/welcome-to-kubernetes-new-developer-guide.md:
```diff
@@ -0,0 +1,126 @@
+[development.md](https://github.com/kubernetes/community/blob/master/contributors/devel/development.md)
+for more details.
+
+Pull-Request Process
```
Rather than re-write the book a second time, the PR process is documented pretty well in pull-requests.md in the same directly. Maybe you can link refer to that?

alexandercampbell replied:
Good call. I've linked directly to `pull-requests.md`.

cblecker commented on contributors/devel/welcome-to-kubernetes-new-developer-guide.md:
```diff
@@ -0,0 +1,126 @@
+Downloading, Building, and Testing Kubernetes
+---------------------------------------------
+
+This guide is non-technical, so I'm going to refer you to our online
```
This should be reworded, as this isn't a personal doc, so it shouldn't speak in first person.

cblecker commented on contributors/devel/welcome-to-kubernetes-new-developer-guide.md:
```diff
@@ -0,0 +1,126 @@
+---------------------------------------------
+
+This guide is non-technical, so I'm going to refer you to our online
+documentation.We have plenty of documentation available under
```
nit: space after period

cblecker commented on contributors/devel/welcome-to-kubernetes-new-developer-guide.md:
```diff
@@ -0,0 +1,126 @@
+[github.com/kubernetes/test-infra](https://github.com/kubernetes/test-infra) and
+their names are:
+
+ 1. @k8s-bot
```
k8s-bot doesn't directly reply to anything atm

cblecker commented on contributors/devel/welcome-to-kubernetes-new-developer-guide.md:
```diff
@@ -0,0 +1,126 @@
+[OWNERS](https://github.com/kubernetes/community/blob/master/contributors/devel/owners.md)
+file of the your changes' root directory and you should get a quicker response.
+
+In order for your PR to be merged, you need four things. When you have all four,
```
This is more clearly detailed in pull-requests.md

cblecker commented on contributors/devel/welcome-to-kubernetes-new-developer-guide.md:
```diff
@@ -0,0 +1,126 @@
+    to rerun the tests.
+
+There is an entire SIG
+([sig-contributor-experience](https://github.com/kubernetes/kubernetes/wiki/CLA-FAQ))
```
Link to the contribex sig doc in k/community instead of the wiki which is out of date.

cblecker commented on contributors/devel/welcome-to-kubernetes-new-developer-guide.md:
```diff
@@ -0,0 +1,126 @@
+freeze usually lasts two weeks and happens once per quarter.
+
+If you're new to Kubernetes, you won't have to worry about this too much. After
+you've contributed for a few months, you will likely [become an official SIG
```
there isn't such thing as "an official sig member". there are org members, and sig leads.. but participation in a sig is basically just showing up :)

cblecker commented on contributors/devel/welcome-to-kubernetes-new-developer-guide.md:
```diff
@@ -0,0 +1,126 @@
+If you're new to Kubernetes, you won't have to worry about this too much. After
+you've contributed for a few months, you will likely [become an official SIG
+member](https://github.com/kubernetes/community/blob/master/community-membership.md)
+and take ownership of some of the release tests. At this point, you'll work with
```
There is test ownership which is sig based, but also release tests which are maintained by sig-release.

cblecker commented on contributors/devel/welcome-to-kubernetes-new-developer-guide.md:
```diff
@@ -0,0 +1,126 @@
+about release milestones. This happens when we are in a code freeze for a
+release. In order to ensure Kubernetes is stable, we stop merging everything
+that's not a bugfix, then focus on making all the release tests pass. This code
+freeze usually lasts two weeks and happens once per quarter.
```
Link to release schedule, as that details when. Also the sig-release channel and kubernetes-dev@ list will detail and announce when this happens.

alexandercampbell replied:
Good idea, but I could not find a release schedule in `community` or by Googling. Where do I find that?

cblecker replied:
the 1.8 schedule is here: https://github.com/kubernetes/features/blob/master/release-1.8/release-1.8.md

cblecker commented on contributors/devel/welcome-to-kubernetes-new-developer-guide.md:
```diff
@@ -0,0 +1,126 @@
+document will help you understand the organization of the Kubernetes project and
+direct you to the best places to get started. By the end of this doc, you'll be
+able to pick up issues, write code to fix them, and get your work reviewed and
+merged.
```
You don't otherwise mention in this doc where to go to ask questions.. developer support. kubernetes-dev@, slack, etc.

cblecker replied:
Oh, also, reviewing and linking to https://github.com/kubernetes/community/tree/master/contributors/devel#kubernetes-developer-guide and potentially updating it with this would be helpful. That doc kind of just lists the pertinent docs in this dir for various topics.

alexandercampbell replied:
Good call. Added a paragraph below this with links to the Slack Channel and the mailing list.

alexandercampbell commented:
@cblecker I'm about halfway through addressing your comments and it seems like most of this information is well-covered elsewhere. Is there another doc that would be better for me to merge this into, rather than making a new document?

cblecker commented:
@alexandercampbell I don't think it's a bad idea to have a clearly marked start point for new contributors. Like even the contributors/devel dir today has 53 markdown docs. But there's no need to duplicate info.. summaries or descriptions of why things are important, and highlighting the other more detailed docs work work just as well.

MHBauer commented on contributors/devel/welcome-to-kubernetes-new-developer-guide.md:
```diff
@@ -0,0 +1,126 @@
+The Continuous Integration (CI) robot, @k8s-ci-robot, won't run the tests until
+a Kubernetes maintainer gives the OK. We require PRs to be approved for testing
+in order to prevent people from running malicious code on our CI servers, for
+example mining for bitcoins.
```
I don't think you need the example.

alexandercampbell commented:
@cblecker @MHBauer thank you for the detailed reviews! I believe I've addressed all of your comments (or responded with questions). The doc is a lot shorter now that I have deleted sections to link to existing documentation instead. How does it look now?

cblecker commented on contributors/devel/welcome-to-kubernetes-new-developer-guide.md:
```diff
@@ -0,0 +1,91 @@
+Pull-Request Process
+--------------------
+
+The pull-request process is documented in [pull-requests.md](pull-requests.md).
```
```
Contributing to Kubernetes is done through pull requests to the various repos. This process is documented in detail [here](pull-requests.md).
```
Reading over this too, the one thing that might be worth adding back here is something about the CLA and linking to `(../../CLA.md)`. It's an important step for brand new contributors, and you really only need to deal with it once.

alexandercampbell replied:
Done :)

alexandercampbell commented:
@cblecker I think I've gotten to all of your comments on the most recent round. Thanks for the help!

cblecker commented:
@alexandercampbell Thanks for all your work here! Feel free to squash what you have. We can leave this 24h for anyone else who wants to review, and then merge :)

alexandercampbell commented:
Sounds good to me. Squashed.

alexandercampbell commented:
@cblecker seems nobody else has comments. Does this still look good?

alexandercampbell commented:
@cblecker ping :)

cblecker commented:
/lgtm
